operator/src/main/java/com/instaclustr/cassandra/operator/service/ControllerService.java

```diff
@@ -1,428 +1,435 @@
+        }
 
-    private static void deleteResource(final ApiCallable deleteCallable) throws ApiException {
-        try {
-            deleteCallable.call();
+        return Integer.valueOf(matcher.group("index"));
+    }).reversed();
 
-        } catch (final ApiException e) {
-            if (e.getCode() != 409)
-                throw e;
+    private void replaceStatefulSet(final DataCenter dataCenter, final V1beta2StatefulSet statefulSet) throws ApiException {
+        final int currentReplicas;
+        {
+            final V1beta2StatefulSet existingStatefulSet = appsApi.readNamespacedStatefulSet(statefulSet.getMetadata().getName(), statefulSet.getMetadata().getNamespace(), null, null, null);
+            currentReplicas = existingStatefulSet.getSpec().getReplicas();
         }
-    }
+        final int desiredReplicas = dataCenter.getSpec().getReplicas();
+
+        // ideally this should use the same selector as the StatefulSet.
+        // why does listNamespacedPod take a string for the selector when V1LabelSelector exists?
+        final String labelSelector = String.format("cassandra-datacenter=%s", dataCenter.getMetadata().getName());
+
+        // TODO: handle lists larger than the max returned (via continue attribute)
+        final List<V1Pod> pods = ImmutableList.sortedCopyOf(STATEFUL_SET_POD_COMPARATOR,
+                coreApi.listNamespacedPod(dataCenter.getMetadata().getNamespace(), null, null, null, null, labelSelector, null, null, null, null).getItems()
+        );
 
-    private void deleteStatefulSet(final V1beta2StatefulSet statefulSet) throws Exception {
-        V1DeleteOptions deleteOptions = new V1DeleteOptions();
-        deleteOptions.setPropagationPolicy("Foreground");
-
-        //Scale the statefulset down to zero (https://github.com/kubernetes/client-go/issues/91)
-        statefulSet.getSpec().setReplicas(0);
-        try {
-            appsApi.replaceNamespacedStatefulSet(statefulSet.getMetadata().getName(), namespace, statefulSet, null);
-        } catch (ApiException e) {
-            throw e;
-        }
-        while (true) {    		
-            Integer currentReplicas = appsApi.readNamespacedStatefulSet(statefulSet.getMetadata().getName(), namespace, null, null, null).getStatus().getReplicas();
-            if (currentReplicas == 0) 
-                break;
 
-            Thread.sleep(50);
+        // TODO: this will be slow for a large number of nodes -- introduce some parallelism
+        final ImmutableListMultimap<CassandraConnection.Status.OperationMode, V1Pod> podCassandraOperationModes = Multimaps.index(pods, p -> {
+            final CassandraConnection cassandraConnection = cassandraConnectionFactory.connectionForPod(p);
+
+            return cassandraConnection.status().operationMode;
+        });
+
+
+        if (EnumSet.of(CassandraConnection.Status.OperationMode.NORMAL, CassandraConnection.Status.OperationMode.DECOMMISSIONED).equals(podCassandraOperationModes.keySet())) {
```

Comment:
Seems if someone tries to scale statefulset down to 0, he can't achieve that because this condition only works for a mix set of normal and decommisioned (might be a invalid ask, my concern is only for that bug I mentioned before, can't think any other scenarios need to do so)